libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt

```diff
@@ -0,0 +1,11 @@
+package com.mapbox.navigation.base.session
+
+interface NavigationSessionObserver {
```

Comment:
Same as above.

Reply:
Same as above

Reply:
I see the usefulness. I'm just worried that by exposing those states, we're exposing internal mechanisms and tying ourselves to certain expectations that could not be changed later on with additional features. After all, we might be in the internal "Free Drive" mode, but if there are no tiles on the device, enhanced location will be equal to raw location and user's shouldn't have to worry about it etc. Same for `Active Guidance`, user can change to the Navigation UI view whenever convenient, we don't have to be in the internal `Active Guidance` state. The states bring value to us (like billing) but might not to the end-user. WDYT @mapbox/navigation-android?

Regardless, these observers can live in `core` only and be exposed from there, no modules depend on them.

Reply:
What if the `MapboxService` or `TripNotification` wants to listen to the state changes? I think the existing telemetry code from legacy SDK also performs some actions based on active navigation. How would `metrics` module which contains `telemetry` in the new v1.0 listen to these navigation state changes?

Reply:
I think the metrics module is being moved to core in https://github.com/mapbox/mapbox-navigation-android/pull/2445. But regardless, if we need to push a state from `MapboxNavigaiton` to an upstream module, it means that the module is probably setup incorrectly because it has dependencies on the downstream logic.

Reply:
Then we need to think about it. Navigation State changes is something that should be available to everyone. The architecture might have to be modified a little bit to address this [issue](https://github.com/mapbox/navigation-sdks/issues/231). We may end up doing some stuff in background which always runs even if the activity is force closed from recents. I know the legacy code in this case needed to know active guidance state in order to decide when to close this background job.
We need to consider how we want to expose the state changes both to the different modules in the SDK and end clients as well

Reply:
Created ticket to explore how we may want to modify it in the next iteration.